cmd/dp-hierarchy-builder/main.go

```diff
@@ -2,127 +2,194 @@ package main
 	errorHandler, err := reporter.NewImportErrorReporter(kafkaErrorProducer, log.Namespace)
-	exitIfError(err)
+	exitIfError(ctx, err, "error creating import error reporter")
 
 	eventHandler := event.NewDataImportCompleteHandler(&hierarchy.Store{db}, avroProducer)
 
 	eventConsumer := event.NewConsumer()
-	eventConsumer.Consume(kafkaConsumer, eventHandler, errorHandler)
-
-	healthChecker := healthcheck.NewServer(
-		cfg.BindAddr,
-		cfg.HealthCheckInterval,
-		cfg.HealthCheckRecoveryInterval,
-		errorChannel,
-		db,
-	)
+	eventConsumer.Consume(ctx, kafkaConsumer, eventHandler, errorHandler)
+
+	hc := startHealthChecks(ctx, cfg, kafkaConsumer, kafkaProducer, kafkaErrorProducer, db)
+
+	apiErrors, httpServer := startApi(ctx, hc, cfg)
 
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
 
 	// this will block (main) until a fatal error occurs
 	select {
-	case err := <-kafkaConsumer.Errors():
-		log.ErrorC("kafka consumer", err, nil)
-	case err := <-kafkaProducer.Errors():
-		log.ErrorC("kafka result producer", err, nil)
-	case err := <-kafkaErrorProducer.Errors():
-		log.ErrorC("kafka error producer", err, nil)
+	case err := <-apiErrors:
+		log.Event(ctx, "http server error", log.ERROR, log.Error(err))
+	case err := <-kafkaConsumer.Channels().Errors:
+		log.Event(ctx, "kafka consumer error", log.ERROR, log.Error(err))
+	case err := <-kafkaProducer.Channels().Errors:
+		log.Event(ctx, "kafka result producer error", log.ERROR, log.Error(err))
+	case err := <-kafkaErrorProducer.Channels().Errors:
+		log.Event(ctx, "kafka error producer error", log.ERROR, log.Error(err))
 	case err := <-errorChannel:
-		log.ErrorC("error channel", err, nil)
+		log.Event(ctx, "error channel error", log.ERROR, log.Error(err))
 	case <-signals:
-		log.Debug("os signal received", nil)
+		log.Event(ctx, "os signal received", log.INFO)
 	}
 
-	ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
+	ctx, cancel := context.WithTimeout(ctx, cfg.GracefulShutdownTimeout)
 
 	// gracefully dispose resources
+	hc.Stop()
+
+	err = httpServer.Close(ctx)
+	logIfError(ctx, err, "error closing http server")
+
 	err = eventConsumer.Close(ctx)
-	logIfError(err)
+	logIfError(ctx, err, "error closing event consumer")
 
 	err = kafkaConsumer.Close(ctx)
-	logIfError(err)
+	logIfError(ctx, err, "error closing kafka consumer")
 
 	err = kafkaProducer.Close(ctx)
-	logIfError(err)
+	logIfError(ctx, err, "error closing kafka producer")
 
 	err = kafkaErrorProducer.Close(ctx)
-	logIfError(err)
-
-	err = healthChecker.Close(ctx)
-	logIfError(err)
+	logIfError(ctx, err, "error closing kafka error producer")
 
 	err = db.Close(ctx)
-	logIfError(err)
+	logIfError(ctx, err, "error closing graph db connection")
 
 	// cancel the timer in the shutdown context
 	cancel()
 
-	log.Debug("graceful shutdown was successful", nil)
+	log.Event(ctx, "graceful shutdown was successful", log.INFO)
```

Comment:
Like your comment on my pr, this may not be a successful graceful shutdown if any of the connections failed to close or cancel was called due to context timing out; see observation importer to resolve this